transaction/StronglyConnectedComponents.hpp

```diff
@@ -0,0 +1,113 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_TRANSACTION_STRONGLY_CONNECTED_COMPONENTS_HPP_
+#define QUICKSTEP_TRANSACTION_STRONGLY_CONNECTED_COMPONENTS_HPP_
+
+#include <stack>
+#include <vector>
+#include <unordered_map>
+
+#include "transaction/DirectedGraph.hpp"
+#include "utility/Macros.hpp"
+
+namespace quickstep {
+
+namespace transaction {
+
+/** \addtogroup Transaction
+ *  @{
+ */
+
+/**
+ * @brief Class for applying Tarjan's strongly connected components
+ *        algorithm to the 
+ **/
+class StronglyConnectedComponents {
+ public:
+  /**
+   * @brief Constructor for StronglyConnectedComponents.
+   *
+   * @param directed_graph Pointer to the directed graph which the 
+   *        algortihm runs.
+   **/
+  explicit StronglyConnectedComponents(DirectedGraph *directed_graph);
+
+  /**
+   * @brief Enumaretes strongly connected components.
+   * @info This method should be called before calling any other method.
+   **/
+  void findStronglyConnectedComponents();
+
+  /**
+   * @brief Getter for the component id of the node.
+   *
+   * @param node_id Id of the node in the graph.
+   * @return Component id of the node.
+   **/
+  std::uint64_t getComponentId(DirectedGraph::NodeId node_id) const;
+
+  /**
+   * @brief Getter for total number of strongly connected components.
+   * @warning This method should be called after findStronglyConnectedComponents().
+   *
+   * @return Total number of strongly connected components.
+   **/
+  std::uint64_t getTotalComponents() const;
+
+  /**
+   * @brief Gets a component id to list of node ids mapping.
+   *
+   * @return Mapping of component ids to list of nodes.
+   **/
+  std::unordered_map<std::uint64_t, std::vector<DirectedGraph::NodeId>> getComponentMapping() const;
+
+ private:
+  // Applies DFS to the node whic has id v.
+  void depthFirstSearch(DirectedGraph::NodeId v);
+
+  // Pointer to the graph.
+  DirectedGraph *directed_graph_;
```

Comment:
Who "owns" the graph that is pointed to. Should make the contract clear, especially if that graph could change between calls to methods here.